v3 without conan

rturrado opened the pull request:
### CMake

Updated C++ version to C++23.
Updated GitHub Actions to build in gcc/clang/msvc x Linux/MacOS/Windows x Debug/Release.
Updated dependency management to be done through FetchContent instead of git submodules.
- CMake minimum required version is now 3.25, due to FetchContent_Declare(... SYSTEM).
- lemon dependency is now fetched through a fork, which fixes C++17 & 20 compilation, and some CMake and C++ warnings.

### Reorganized tests, examples, and resources.

Reorganized tests.
- All tests are now under the `test` folder.
- All tests are `googletest`. `doctest` dependency has been removed.
- All tests are discovered. CMake code for adding tests has been removed.
- Python tests have been updated to comply to most of PEP 8 (code style guide).

Reorganized resources.
- All resources are now under the `res` folder.
- Some file names were changed.

Reorganized examples.
- All example codes are now under `example` folders within `test`.
- All example resources are now under `example` folders within `res`.

The structure of the `test` folder is:
```
  |- ql
  |  \- com, ir, pass, utils...
  \- v1x
     |- cpp
     |  \- example
     \- python
        |- example
        \- visualizer
```
The structure of the `res/v1x` folder is:
```
  |- cq
  |  |- example
  |  \- golden
  |- json
  |  \- visualizer
  \- qasm
     \- golden
```
`v1x` stands for cQasm v1.x. There may be folders such as `v3x` in the future.

### File changes

- deps: removed.
- docs:
  - Changed `from openql import openql as ql` to `import openql as ql`.
  - Changed `nqubits` to `num_qubits`.
- include/ql, source/ql:
  - Fixed warnings and compilation errors.
  - Update references to resources to point to the new location in `res`.
- python/ql.i: minor aesthetic change: split a very whole string into different lines.
- .gitmodules: removed.
- CMakeLists.txt: C++23, FetchContent, `ql` target_compile_options.
- pytest.ini: updated testpaths to point to `test/v1x/python`.

rturrado committed:
Initial commit.
Added a possible test folder structure.
Added an empty test/ql/ir/cqasm/read.cc test file.

rturrado committed:
Implemented some parse version tests.
read.cc: moved parse code to a separate function, which is the one that is being tested.

rturrado committed:
include/ql/ir/cqasm/read.h:
- renamed 'read' as read_v1 and added read_v3.
src/ql/ir/cqasm/read.cc:
- 'read' now parses the program version and calls whether read_v1 or read_v3.
- read_v3 still doesn't do anything.
test/ql/ir/cqasm/read.cc:
- commented out tests for versions older than 1.2 as they need an IR with just more than a 'version' statement.
- tests now create an IR and a basic platform and call 'read', passing a very simple program that only has a 'version' statement.

rturrado committed:
test/ql/ir/cqasm/read.cc: added several fake structures, the main one being PlatformFake. Now the test is simplified a bit.

rturrado committed:
Fixed some compilation warnings.

rturrado committed:
Added libqasm/v3 via FetchContent.
Added OpenQL/v3 to GitHub Actions.

rturrado committed:
Fixed failing test.

rturrado committed:
Fix typo. [skip ci]

rturrado committed:
Merge remote-tracking branch 'origin/v3' into v3

rturrado committed:
Changed OpenQL C++ standard to C++20.
This change makes some of the git submodules to stop working, since they are not C++20 compliant.
That doesn't mean we can not still use their binaries.
Updated GitHub workflow's cpp job to use Conan.
Updated README file.

.gitmodules, CMakeLists.txt, conanfile.py:
- Removed all git submodules.
- Added backward-cpp, cimg, eigen, highs, nlohmann_json, doctest and gtest as Conan dependencies.
- libqasm and lemon are now obtained via FetchContent, since there are not yet Conan packages for them.
- lemon is patched after download to fix C++17 and C++20 compilation warnings.

circuit.h/cc, functions.h/cc, gate.h, list.h, mappings.h, old_to_new.cc, past.cc, prim.h, types.h, unitary.cc, vec.h: fixed C++20 compilation warnings and errors.

rturrado committed:
Disabled OPENQL_BUILD_TESTS for the time being.
TODO: organize tests in OpenQL as we did for libqasm.
Updated setup.py so that it calls Conan instead of CMake.
Added a new patch for lemon.
Removed highs from .gitmodules.

rturrado committed:
Amend of last commit.

rturrado committed:
Trying libqasm/v3_distro.

rturrado committed:
Merge remote-tracking branch 'origin/v3' into v3

rturrado committed:
CMakeLists.txt: back to libqasm/v3, since OpenQL 1) requires tree-gen anyway, and 2) and most importantly, since libqasm is fetched via FetchContent, it will require flex/bison/m4/tree-gen for its compilation.
conanfile.py: added flex/bison/m4 as dependencies to conanfile.py.
test.yml: commented out everything except for C++ Windows builds.

rturrado committed:
Fixed C++20 Windows compilation.

rturrado committed:
Forgot to add patches/adaptors.h.

rturrado committed:
Fixed ambiguities in operator== introduced by changes in C++20.
C++20 adds a 'synthesized' y==x for each x==y, what can lead to ambiguities in existing C++17 code.
In our case, many of these operator== implementations come from code generated by tree-gen.
A possible solution would be to add friend operator== functions to resolve the ambiguities.
A probably less elegant solution, that is what I've done here, and that avoids touching the code generated by tree-gen, is to use the x.operator==(y) syntax in the caller side.

rturrado committed:
CMakeLists.txt: reverted last change as it introduces other errors.
I've finally added a non-member operator==(const Object &, const PhysicalObject &) to resolve the code ambiguities due to moving to C++20:
- We leave all the existing code untouched, and using the simpler syntax 'x == y' instead of 'x.operator==(y)'.
- We don't change anything either in ir.tree (e.g. adding the non-member function there as a friend to PhysicalObject, for example).
- We simply add the function in 'ir_gen_ex.cc' (read it as an extension to ir.gen.cc, the file generated by tree-gen).

rturrado committed:
ir_gen_ex.cc: changed free function operator== signature.
test.yml: going to test only macos for the moment.

rturrado committed:
Testing C++23.

rturrado committed:
Testing C++23 Windows compilation.

rturrado committed:
tst.yml: fixed 'conan build' line.
Testing C++23 Ubuntu/MacOS/Windows compilation.

rturrado committed:
Files with calls to ambiguous operator== should see free function operator==(const Object&, const PhysicalObject&).

rturrado committed:
Files with calls to ambiguous operator== should see free function operator==(const Link<Object>&, const Link<PhysicalObjec>t&).

rturrado committed:
operands.cc: calls to ambiguous operator== should see free function operator==(const Link<Object>&, const Link<PhysicalObjec>t&).

rturrado committed:
Added free function operator==(const Link<Object>&, const Link<Object>&).

rturrado committed:
lemon/lfgw_writer.h: trying to fix macos compilation.

rturrado committed:
conanfile.py: added verbosity to conan build.
CMakeLists.txt: added a debug comment to the patch_lemon_sources_command.
test.yml: testing only MacOS.

rturrado committed:
CMakeLists.txt: changed the copy commands and printed the lfg_writer.h after the copy.

rturrado committed:
Renamed lfg_writer.h to lgf_writer.h.

rturrado committed:
Added lgf_reader.h to the list of lemon patches.
MacOS builds cannot find a viable constructor when returning a SectionReader from a function.
What I understood from the error messages was that an rvalue to a SectionReader was being passed to a constructor, and, thus, that a move constructor was being needed.
That's why I changed the SectionReader(SectionReader&) constructor into a SectionReader(SectionReader&&).

rturrado committed:
At this point, the CMake's configure step fails because both eigen and lemon have a same target called 'check'.
I would need to know how this is not a problem in current OpenQL.
It may be due to the fact that I'm using eigen's GitHub version, whereas OpenQL's git submodule uses the GitLab version (although they clain that the GitHub website is an exact mirror of the other one).
One way to go further would be to create a QuTech-Delft fork of the lemon library, where we could change the name of the 'check' target, apart from committing the files we currently have in the patches directory (and that fix the C++17/20 compilation).

rturrado committed:
Fixed CMakeLists.txt:
- Changed Eigen to the gitlab version.
- Changed lemon to rturrado's fork version.
- Updated ql's include directories.
- Updated ql's link libraries.

rturrado committed:
OpenQL v3 candidate for PR:
- Using C++23.
- Using latest libqasm/v3 (PR approved to merge to develop).
- Managing dependencies via FetchContent, not git submodules.
- Passing Continuous Integration.
CMakeLists.txt:
- All dependencies are managed through FetchContent.
- Enabled tests, and added test directory.
- Added policies 48 and 77.
- Updated GIT_TAG for cqasm, Eigen3 (master), and lemon (rturrado's fork).
- Added OPENQL_BUILD_CPP_STANDALONE_OPTION.
- Added ${BACKWARD_ENABLE} to add_library(ql) and add_backward(ql).
- Added some debug messages.
- We don't hate CMake anymore.
setup.py:
- Removed conan code.
- Commented in tests and installation blocks.
test/CMakeLists.txt:
- Changed source management to use target_sources.
test.yml:
- cpp job now builds the standalone example and runs it.
- Added OPENQL_DISABLE_UNITARY=true to python job.
- Removed cpp-standalone job.
- Removed conda code, as it was commented out.
src/ql/ir/cqasm:
- Removed references to cqasm::v3, and read_v3 function.
test/ql/ir/cqasm/read.cc:
- Updated tests.
Other:
- Removed .gitmodules.
- Removed conanfile.py.
- Removed examples/cpp_standalone_example/CMakeLists.txt.
- Removed patches folder.

rturrado committed:
Trying to fix Eigen3 CMake configuration:
-  It is failing in GitHub Actions, saying "In-source builds not allowed. Please make a new directory (called a build directory) and run CMake from there."
Also updated actions versions in test.yml.

rturrado committed:
test/CMakeLists.txt: fixed Windows build.
test.yml: OpenQL_test is already run by ctest.

rturrado committed:
test.yml: run './example' for non-Windows systems, and 'example.exe' for Windows systems.

rturrado committed:
test.yml: run './example' for non-Windows systems, and not './example.exe'.

rturrado committed:
Fixed Clang compilation.
Testing with a set of test.yml jobs similar to those of QX-simulator.
CMakeLists.txt:
- Updated cqasm version.
- Changed highs linkage to PRIVATE
- Changed Clang compile options so that some warnings are not treated as errors.

rturrado committed:
Fixed Visualizer errors (after enabling X11 locally).
CMakeLists.txt:
- Fixed '<lowercaseName_SOURCE_DIR>' references.
- Added extra errors to gcc and Clang compilation.
- Changed CImg to be fetched only when needed.
test.yml: fixed cpp-arm64 job.

rturrado committed:
CMakeLists.txt:
- Added 'Wno-error=unused-private-field' to Clang compile options.
test.yml:
- Fixed cpp-linux and cpp-windows jobs.
unitary.cc:
- Fixed unused parameter errors.

rturrado committed:
CMakeLists.txt: temporarily commented out 'add_subdirectory(test)' to check if that fixes the memory exhaustion issues when building gcc/Linux/x64.
kernel.cc: trying to fix gcc/Linux/ARM64 build.
unitary.cc: fixed to_string method (Python builds).

rturrado committed:
Fixed errors of the style 'return "blah" + var;' and 'var += "blah" + other_var;' (ARM64 builds)
CMakeLists.txt:
- Fixed management of highs dependency.
- Updated lemon library (to a version that removes one compilation warning, and sets a CMake policy 48 to new).
- Added 'Wno-error=deprecated-declarations' (Python ubuntu builds).
- 'add_subdirectory(test)' commented in again.
test.yml:
- Trying to fix Windows builds.

rturrado committed:
CMakeLists.txt:
- Added 'SYSTEM' to 'add_subdirectory' for highs, so that warnings in highs code are not treated as errors in our project.
test.yml:
- Fixed way of executing 'example.exe' in Windows.
- Fixed '$env:GITHUB_ENV' references.

rturrado committed:
CMakeLists.txt:
- Set CMake minimum required version to 3.25.
- Moved FetchContent code after global configuration.
test.yml:
- Added 'Get latest CMake' step to all jobs.

rturrado committed:
CMakeLists.txt:
- Added 'Wno-error=restrict' Clang compile option.
test.yml:
- Set 'LIBQASM_BUILD_TYPE=Debug' and 'OPENQL_DISABLE_UNITARY=true' for Python/ubuntu job.
Reverted previous fixes turning "blah" into std::string{"blah"}.

rturrado committed:
CMakeLists.txt: moved 'Wno-error=restrict' from Clang to gcc compile options.

rturrado committed:
Propagate ql library compile options to tests.

rturrado committed:
test.yml: fixed cpp-windows and cpp-arm64 jobs.
CMakeLists.txt: set ql compile options back to PRIVATE and add 'Wno-error=restrict' to tests' compile options.

rturrado committed:
Set the same compile options for the test executables than for the ql library.

rturrado committed:
CMakeLists.txt:
- Temporarily commented out 'add_subdirectory(test)'.
- Minor aesthetic changes to the FetchContent sections.
test.yml:
- Removed OPENQL_BUILD_CPP_STANDALONE_EXAMPLE option, and 'example' execution from all the jbs.
- Changed '$env::GITHUB_ENV' to '$GITHUB_ENV' for Python Windows builds.

rturrado committed:
Testing a possible fix for memory exhaustion issues.
CMakeLists.txt:
- 'add_subdirectory(test)' commented in again.
test.yml:
- Changed number of parallel build processes from 10 to 5.
- Changed '$GITHUB_ENV' back to '$env:GIHTUB_ENV' for Python Windows builds.

rturrado committed:
Reorganized tests.
- All tests are now under the 'test' folder.
- All tests are googletest. doctest dependency has been removed.
- All tests are discovered. CMake code for adding tests has been removed.
- Python tests have been updated to comply to most of PEP 8 (code style guide).

Reorganized resources.
- All resources are now under the 'res' folder.

Reorganized examples.
- All example codes are now under 'example' folders within 'test'.
- All example resources are now under 'example' folders within 'res'.

The structure of the 'test' folder is:
  |- ql
  |  \- com, ir, pass, utils...
  |- v1x
  |  |- cpp
  |  |  \- example
  |  |- python
  |     |- example
  |     \- visualizer

The structure of the 'res/v1x' folder is:
  |- cq
  |  |- example
  |  \- golden
  |- json
  |  \- visualizer
  |- qasm
     \- golden

'v1x' stands for cQasm v1.x. There may be folders such as 'v3x' in the future.

rturrado committed:
Merge remote-tracking branch 'origin/v3_without_conan' into v3_without_conan

# Conflicts:
#	CMakeLists.txt

rturrado committed:
Fixed cpp tests (over which I had done some wrong search/replace).

rturrado committed:
CMakeLists.txt: updated cqasm to the latest libqasm/develop, which includes all the libqasm/v3 work.
Updated v1 references to v1x.

rturrado committed:
CMakeLists.txt: avoid using FetchContent_Populate.

rturrado committed:
test.yml: changed 'on-push-branches' from 'v3_without_coman' to 'develop' prior to the pull request. [skip ci]

rturrado committed:
Merge branch 'v3_without_conan' into develop

# Conflicts:
#	.github/workflows/test.yml
#	include/ql/ir/compat/gate.h
#	res/v1x/json/debug.cc
#	res/v1x/json/test_cfg_none_simple.json
#	res/v1x/json/test_multi_core_64x16_full.json
#	res/v1x/json/test_unitary.cc
#	res/v1x/qasm/golden/test_condex_toffoli_compos_gate_last.qasm
#	res/v1x/qasm/golden/test_mc_non_comms_last.qasm
#	res/v1x/qasm/test_gate_decomposition_cz.cc_backend.map
#	res/v1x/qasm/test_gate_decomposition_cz.cc_backend.vcd
#	res/v1x/qasm/test_gate_decomposition_cz.cc_backend.vq1asm
#	res/v1x/qasm/test_gate_decomposition_cz.initial.cq
#	res/v1x/qasm/test_gate_decomposition_cz.scheduled.cq
#	res/v1x/qasm/test_qi_example.cc_backend.map
#	res/v1x/qasm/test_qi_example.cc_backend.vcd
#	res/v1x/qasm/test_qi_example.cc_backend.vq1asm
#	res/v1x/qasm/test_qi_example.initial.cq
#	res/v1x/qasm/test_qi_example.scheduled.cq
#	src/ql/arch/cc/pass/gen/vq1asm/detail/backend.cc
#	src/ql/arch/cc/pass/gen/vq1asm/detail/functions.cc
#	src/ql/ir/new_to_old.cc
#	src/ql/pass/ana/visualize/detail/types.h
#	src/ql/pass/map/qubits/map/detail/alter.cc
#	src/ql/pass/map/qubits/map/detail/free_cycle.cc
#	src/ql/pass/map/qubits/map/detail/mapper.cc
#	src/ql/pass/map/qubits/map/detail/past.cc
#	src/ql/pass/map/qubits/place_mip/place_mip.cc
#	tests/CMakeLists.txt
#	tests/test_mapper.py
#	tests/test_unitary.py

rturrado committed:
Fixed conflicts when merging v3_without_conan to develop.

rturrado committed:
.gitignore: updated 'tests' references to point to 'test/v1x/python'.
README.md: removed Dependencies and Installation sections (reverted PR changes).
include/ql/ir/cqasm: removed '#include "cqasm.hpp"', and 'read_v3' function declaration.
test.yml: added 'on/push/pull_request'.
tests: removed.

rturrado committed:
Trying to fix clang and mvsvc bugs.

rturrado committed:
Fixed some more clang errors.

rturrado committed:
Trying parallel build of 10 processors.

rturrado committed:
Options struct at src/ql/pass/map/qubits/place_mip/detail/impl.h sets 'write_model_to_file' to true by default.

The IpGridTest.grid__find_complex_permutation ends up calling Impl::run and writing an 'LpMatrix.pbm' file to the root directory.

We have changed it to write the file to the temporary directory instead.

Another option would be not to write out any file during testing.

rturrado committed:
Changed 'write_model_to_file' member of Options struct to false by default.

rturrado committed:
Revert "Options struct at src/ql/pass/map/qubits/place_mip/detail/impl.h sets 'write_model_to_file' to true by default."

This reverts commit 23f2d1ce8a200ec634daf807de9d6ab8819bf839.

rturrado committed:
Minor tidying up of test files.
Renaming of some resource files.

rturrado committed:
Fixed some Python tests that start failing after renaming some resource files.

rturrado committed:
Updated docs.

rturrado committed:
Fixed IpGridTest::grid__find_complex_permutation.

pablolh commented on .github/workflows/test.yml:
```diff
@@ -7,59 +7,140 @@ on:
-      run: cd tests && ../cbuild/example
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
```
fetch-depth

pablolh replied:
Also think of updating assets.yml

rturrado replied:
Done! Well spotted!

I haven't tested `assets.yml` for none of my last changes in libqasm, OpenQL, and QX-simulator. I guess we'll address possible issues there when we have to cross those bridges.

pablolh commented on .github/workflows/test.yml:
```diff
@@ -68,109 +149,44 @@ jobs:
+          fetch-depth: 0
+      - uses: actions/setup-python@v4
+        with:
+          python-version: "3.11"
```
you're removing testing for 4 versions of Python
what's the reasoning?
Personally I'm fine with that, but then do we no longer support Python < 3.11?

rturrado replied:
Yes, sorry about this. We can put them back. I used the QX-simulator's `test.yml` as a template, and I saw that only version `3.11` was used there.

pablolh replied:
I think it's fair

pablolh commented on README.md:
```diff
@@ -17,4 +17,4 @@ can be simulated on the QX simulator.
 OpenQL's source code is released under the Apache 2.0 license.
 
 For detailed user and contributor documentation, visit the
```
just noticed the extra "the"

rturrado replied:
Haha, that was a hidden gem! 😄

pablolh commented on setup.py:
```diff
@@ -192,7 +192,6 @@ def run(self):
                     cmd = cmd['--parallel'][nprocs]
                 elif not sys.platform.startswith('win'):
                     cmd = cmd['--']['-j'][nprocs]
-            cmd & FG
```
Is that why CMake was running twice every time I was building the Python package? Good fix!

rturrado replied:
Wow. I thought clang-tidy had spotted that line as "never executing", but now I see the code, I don't see why it wouldn't be executing.

pablolh commented on src/ql/ir/compat/detail/cqasm_reader.h:
```diff
@@ -19,12 +19,12 @@ namespace detail {
-namespace lqi { using namespace ::cqasm::instruction; }
-namespace lqa { using namespace ::cqasm::analyzer; }
-namespace lqp { using namespace ::cqasm::parser; }
+namespace lqs { using namespace ::cqasm ::v1x::semantic; }
```
spurious space?

rturrado replied:
Wow! I wonder how that code was working! 😮

pablolh commented on src/ql/utils/vcd.cc:
```diff
@@ -83,7 +83,7 @@ void Vcd::change(Int var, Int timestamp, const Str &value)
 }
 
 
-void Vcd::change(Int var, Int timestamp, Int value) {
```
I understand that the variable was not used, however here the variable names are clearly useful to know what the intent of the code was... or would have been...
This class is not inherited, so here is the only way to know what the code was...

This code has no owner anymore unfortunately. I would suggest not touching it.

rturrado replied:
This is a good comment. I should have, in general, just commented out the variable names.

I've put the variable names back in again for this particular case. They're commented out though to avoid compilation warnings/errors.

pablolh commented on test/ql/com/ddg/ddg.cc:
```diff
@@ -1,13 +1,16 @@
-#include "ql/ir/compat/compat.h"
-#include "ql/ir/old_to_new.h"
 #include "ql/com/ddg/build.h"
-#include "ql/com/ddg/ops.h"
```
I think the intent of placing tests in src was to distinguish between
- unit tests, or so-called unit tests, accompanying the source
- integration tests, in tests/ folder
which is a pretty reasonable organization to me. Unfrotunately in OpenQL it's not well done, there are a bit of integration tests everywhere and almost no unit test, which is very annoying to know the source of an issue and the intent of the code.

Normally the unit tests link against only the piece they test.
OpenQL is just a massive CMake library so it doesn't bring any benefit.
Unfortunately my CMake change didn't go through, see https://github.com/QuTech-Delft/OpenQL/pull/460 where I tried to split CMake targets.

pablolh replied:
If we had smaller CMake targets, having multiple test executables wouldn't be an issue, I think.

rturrado replied:
I'll have to dedicate some time to go through the comments in that other PR! 😄 

We could perfectly try to not mix unit and integration tests in the same binary. That's perfectly sensible. Having, for example, an `OpenQL_unit_test` and an `OpenQL_integration_test`.

What I'm more in favour of, maybe because I think it's more an standard way, is to separate tests from source code. But, again, I may be very wrong here.

I'll leave this conversation unresolved so that we can talk about it next week.

pablolh replied:
I think with CMake it's indeed standard to separate source and tests, fair point.
Let's not discuss the tests problem too much in this PR anyway - it's devoted to sth else.

I'm OK with them moving, actually your tests/CMakeLists.txt is a lot cleaner than the previous testing section in the CMakeLists at the root

pablolh commented on test/v1x/cpp/example/cpp_standalone.cc:
```diff
@@ -0,0 +1,31 @@
+#include <iostream>
+
+
+int main(int argc, char **argv) {
```
I think this could be a google test as the others.
It's an example but can totally be a test in the big test executable.

pablolh replied:
Will also save on link time

rturrado replied:
My idea was to leave it as an example. I think it wasn't built at the moment. But I see no reason either to add it as a test, yes.

pablolh commented:
few comments I think this can be merged quickly!

rturrado committed:
Fixed IpStarTest::star_with_2q_gate__no_perfect_solution.

rturrado committed:
Addressed most of comments from Pablo's code review.
test.yml: removed 'fetch-depth: 0' and moved 'checkout' stop after installation steps and right before creation of the build directory.
assets.yml:
- Updated 'checkout' step to not use 'recursive'.
- Updated GitHub Actions versions.
- Changed manylinux job's 'Install cmake' step to use 'Get latest CMake', as in other jobs.
- Removed dependency to Python 3.7.
- Removed commented code.
src/*.cc: added back in the names of the unused parameters, although commmented out.
cpp_standalone.cc: changed into a test.
